src/main/java/com/yelp/nrtsearch/server/luceneserver/SearchHandler.java

```diff
@@ -176,6 +177,14 @@ public SearchResponse handle(IndexState indexState, SearchRequest searchRequest)
         hits = searcherResult.getTopDocs();
       }
 
+      List<Explanation> explanations = null;
+      if (searchRequest.getExplain()) {
+        explanations = new ArrayList<>();
+        for (int i = 0; i < hits.scoreDocs.length; i++) {
+          explanations.add(s.searcher.explain(searchContext.getQuery(), i));
+        }
```

Comment:
This can be faster if the explanations are created for each document in parallel. The fetchFields method already handles some parallelism at document and field level and this can be added there as well. The downside is that this becomes more complicated as you would need both serial and parallel implementations but the performance improvement especially if large number of documents are retrieved might be worth it.